### PR TITLE
[FIX] mail: reintroduce auto scroll on receiving message

### PR DESCRIPTION
**Pull request by seb-odoo**
task-2336251

### CODE REVIEW

File: addons/mail/static/src/models/message/message.js [hunk 1]
```diff
@@ -300,6 +300,17 @@ function factory(dependencies) {
...
+         * Notifies threadViews about this message being just received. This can
+         * allow them adjust their scroll position if applicable.
+         */
+        notifyThreadViewsMessageReceived() {
+            for (const thread of this.threads) {
+                for (const threadView of thread.threadViews) {
+                    threadView.addComponentHint('message-received', { message: this });
+                }
+            }
+        }
```
Comment (alexkuhn) [inline addons/mail/static/src/models/message/message.js:313]: This feels inappropriate as a public method of `mail.message`.
This should stay private concern of `mail.messaging_notification_handler`.

File: addons/mail/static/src/components/message_list/message_list.js [hunk 2]
```diff
@@ -537,6 +552,12 @@ class MessageList extends Component {
             scrollTop,
             threadViewer: this.threadView.threadViewer,
         });
+        // Automatically scroll to new received messages only when the list is
+        // currently fully scrolled.
+        const hasAutoScrollOnMessageReceived = (this.props.order === 'asc')
+            ? scrollTop === this.el.scrollHeight - this.el.clientHeight
+            : scrollTop === 0;
+        this.threadView.update({ hasAutoScrollOnMessageReceived });
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message_list/message_list.js:560]: Being precisely at top/bottom is not great. Old code had offset of 20 pixels to improve that.

Though the better code should have new message auto-scroll based on (partial) visibility of the last message before reception of message. As long as this decision is made at component-level, it cannot be much improved.

File: addons/mail/static/src/models/composer/composer.js [hunk 3]
```diff
@@ -291,7 +291,8 @@ function factory(dependencies) {
                 thread.loadNewMessages();
             }
             for (const threadView of this.thread.threadViews) {
-                threadView.addComponentHint('current-partner-just-posted-message', { messageId });
+                // Reset auto scroll to be able to see the newly posted message.
+                threadView.update({ hasAutoScrollOnMessageReceived: true });
```
Comment (alexkuhn) [inline addons/mail/static/src/models/composer/composer.js:295]: Potentially race condition between this and the update on scroll in the message list component.

Rarely happens in practice, and not a super big deal.
Acceptable until code in component is moved to models eventually.

File: addons/mail/static/src/models/thread_view/thread_view.js [hunk 4]
```diff
@@ -197,6 +201,11 @@ function factory(dependencies) {
         isPreparingLoading: attr({
             default: false,
         }),
+        /**
+         * Determines whether `this` should automatically scroll on receiving
+         * a new message.
+         */
+        hasAutoScrollOnMessageReceived: attr(),
```
Comment (alexkuhn) [inline addons/mail/static/src/models/thread_view/thread_view.js:208]: Would clarify that reception of new message is detected by means of component hints.

**Review by alexkuhn — CHANGES_REQUESTED** [review]

File: addons/mail/static/src/components/message_list/message_list.js (hunk 2 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/message_list/message_list.js:560]: I think the twitch approach is best to this: there should be a visual to indicate whether auto-scroll is enabled or not "restore auto scroll" which goes to the end of list immediately + restore auto feature.

In this context I think being exactly at the bottom is fine, I wouldn't change that here, though if there is a strong argument its just as easy to just add some extra px in those conditions.

File: addons/mail/static/src/models/composer/composer.js (hunk 3 above)
Comment (seb-odoo) [reply addons/mail/static/src/models/composer/composer.js:295]: I think I had it while testing. But `throttle` on scroll is the main problem for this feature.

Another possible "manual" race condition is to scroll after just posting message and before received bus notif. In this case auto scroll will be disabled but I think the desired result (user decided to kill the scroll feature itself by scrolling again).

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

**Comment by alexkuhn** [comment]
@robodoo r-
red runbot

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js [hunk 5]
```diff
@@ -759,6 +761,20 @@ function factory(dependencies) {
...
 
+        /**
+         * Notifies threadViews about the given message being just received.
+         * This can allow them adjust their scroll position if applicable.
+         *
+         * @private
+         * @param {mail.message}
+         */
+        _notifyThreadViewsMessageReceived(message) {
+            for (const thread of this.threads) {
```
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:772]: `message.threads`

File: addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js [hunk 6]
```diff
@@ -759,6 +761,20 @@ function factory(dependencies) {
...
+                for (const threadView of thread.threadViews) {
+                    threadView.addComponentHint('message-received', { message });
+                }
+            }
+        }
```
Comment (alexkuhn) [inline addons/mail/static/src/models/messaging_notification_handler/messaging_notification_handler.js:777]: newline after

**Review by alexkuhn — APPROVED** [review]
@robodoo r+

**Comment by alexkuhn** [comment]
@robodoo rebase-ff

**Comment by seb-odoo** [comment]
@xdu-odoo my PR that makes separator disappearing to quickly noticeable (see merge error above).

File: addons/mail/static/src/components/message_list/message_list.js [hunk 7]
```diff
@@ -382,8 +371,34 @@ class MessageList extends Component {
...
+            this.threadView.markComponentHintProcessed(hint);
+            return;
+        }
+        if (!this.threadView.hasAutoScrollOnMessageReceived) {
+            this.threadView.markComponentHintProcessed(hint);
+            return;
+        }
+        if (
+            this.threadView.lastVisibleMessage
+            && (message.id < this.threadView.lastVisibleMessage.id)
```
Comment (alexkuhn) [inline addons/mail/static/src/components/message_list/message_list.js:396]: `&&` at end of line

**Review by alexkuhn — APPROVED** [review]
rest looks good

File: addons/mail/static/src/components/message_list/message_list.js (hunk 7 above)
Comment (seb-odoo) [reply addons/mail/static/src/components/message_list/message_list.js]: ok

**Comment by alexkuhn** [comment]
@robodoo r+

**Review by seb-odoo — COMMENTED** [review]
Added small margin of error when checking for scroll position

@robodoo r+

**Comment by seb-odoo** [comment]
Added proper `afterEvent` for problematic scroll test, hopefully this will be good now.

@robodoo r+

**Pull request closed without merging**